shardus_net/src/lib.rs

```diff
@@ -322,51 +322,42 @@ pub fn multi_send_with_header(mut cx: FunctionContext) -> JsResult<JsUndefined>
 
     let data = data_js_string.into_bytes().to_vec();
 
-    // Create oneshot channels for each host-port pair
-    let mut senders = Vec::with_capacity(hosts.len());
-    let mut receivers = Vec::with_capacity(hosts.len());
+    let complete_cb = Arc::new(complete_cb);
+    let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel::<SendResult>();
 
-    // should a check be added to see if ports.len == hosts.len
-    for _ in 0..hosts.len() {
-        let (sender, receiver) = oneshot::channel::<SendResult>();
-        senders.push(sender);
-        receivers.push(receiver);
-    }
+    RUNTIME.spawn(async move {
 
-    let complete_cb = Arc::new(complete_cb);
-    let this = Arc::new(this);
-
-    // Handle the responses asynchronously
-    for receiver in receivers {
-        let channel = channel.clone();
-        let complete_cb = complete_cb.clone();
-        let this = this.clone();
-
-        RUNTIME.spawn(async move {
-            let result = receiver.await.expect("Complete send tx dropped before notify");
-
-            if await_processing {
-                RUNTIME.spawn_blocking(move || {
-                    channel.send(move |mut cx| {
-                        let cx = &mut cx;
-                        let stats = this.to_inner(cx).get::<JsBox<RefCell<Stats>>, _, _>(cx, "_stats")?;
-                        (**stats).borrow_mut().decrement_outstanding_sends();
-
-                        let this = cx.undefined();
-
-                        if let Err(err) = result {
-                            let error = cx.string(format!("{:?}", err));
-                            complete_cb.to_inner(cx).call(cx, this, [error.upcast()])?;
-                        } else {
-                            complete_cb.to_inner(cx).call(cx, this, [])?;
-                        }
-
-                        Ok(())
-                    });
-                });
-            }
-        });
-    }
+        let mut results = Vec::new();
+
+        // recv will return None when all tx are dropped
+        // So this'll not hang forever.
+        while let Some(result) = rx.recv().await {
+            results.push(result);
+        }
+
+        if schedule_complete_callback {
+            nodejs_thread_channel.send(move |mut cx| {
+                let cx = &mut cx;
+                let stats = this.to_inner(cx).get::<JsBox<RefCell<Stats>>, _, _>(cx, "_stats")?;
+                (**stats).borrow_mut().decrement_outstanding_sends();
```

Comment:
Since we are incrementing the stats counter by N we need to decrement here by N

Reply:
Agreed pushed a new commit to address this.